### PR TITLE
Huggingface_SB3 v2.0 

### PR DESCRIPTION
**Pull request by simoninithomas**
👋 so here's the SB3 v2.0:

With our new version we can use `package_to_hub` method that:
1. Save the model
2. Evaluate the model and generate a results.json
3. Generate a model card
4. Record a replay video of the agent
5. Push everything to the hub
Here's an example : https://huggingface.co/ThomasSimonini/TEST2-Colab-ppo-LunarLander-v2 (very small training so the agent is bad)

If you want to try some examples directly on colab I've made a small test colab: https://colab.research.google.com/drive/1FhZ1w7smqPo8GQcW5qb2HmkggZVuok57?usp=sharing

The PyPi update is also automated thanks to @osanseviero 

A big thanks to Omar who made a lot of tests with the library

I **need a little bit of feedback for the documentation, I think it's not very clear**.
- I explain the 2 cases: downstream and upstream
- In case 3-4 I explain how to use xvfb if you use colab or vm (because you don't have a screen to render so you can't generate a video without xvfb).

WDYT? Thanks

### CODE REVIEW

File: .idea/huggingface_sb3.iml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (osanseviero) [inline .idea/huggingface_sb3.iml:1]: I think things under `.idea` come from IntelliJ. Same for `inspectionProfiles`. I would remove those dirs

File: README.md [hunk 2]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
-Then:
+Then
+
+**With `package_to_hub()`**:
+
+```python
+import gym
+
+from stable_baselines3 import PPO
+from stable_baselines3.common.evaluation import evaluate_policy
```
Comment (osanseviero) [inline README.md:63]: I don't think you need `evaluate_policy` here

File: README.md [hunk 3]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
+from stable_baselines3.common.vec_env import VecVideoRecorder, DummyVecEnv
```
Comment (osanseviero) [inline README.md:64]: I don't think you need `VecVideoRecorder` here

File: README.md [hunk 4]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
+from huggingface_sb3_test import load_from_hub, package_to_hub
```
Comment (osanseviero) [inline README.md:65]: You're not using `load_from_hub` I think

File: README.md [hunk 5]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
+
+# Train the agent
+model.learn(total_timesteps=int(5000))
+
+# This method save, evaluate, generate a model card and record a replay video of your agent before pushing the repo to the hub
+package_to_hub(model=model, 
+               model_name="ppo-LunarLander-v2",
+               model_architecture="PPO",
+               env_id=env_id,
+               eval_env = eval_env,
```
Comment (osanseviero) [inline README.md:85]: nit
```suggestion
               eval_env=eval_env,
```

File: setup.py [hunk 6]
```diff
@@ -2,12 +2,13 @@
 
 install_requires = [
     "huggingface_hub",
-    "cloudpickle==1.6"
+    "cloudpickle==1.6",
+    "pickle5"
 ]
```
Comment (osanseviero) [inline setup.py:7]: I would suggest adding some extras for code style with black, isort, and flake8. This helps follow cleaner code.

See https://github.com/huggingface/huggingface_hub/blob/main/setup.py#L38-L42 as a pointer. If you add that and you do `make style` and `make quality` it will do all of these runs.

File: README.md [hunk 7]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
+               repo_id="ThomasSimonini/TEST2-Colab-ppo-LunarLander-v2",
+               commit_message="Test commit")
+```
+
+
+**With `push_to_hub()`**:
```
Comment (osanseviero) [inline README.md:91]: Maybe here add some explanation that this will only upload the `zip` file.

File: README.md [hunk 8]
```diff
@@ -47,13 +52,52 @@ notebook_login()
...
 ```python
 from huggingface_sb3 import push_to_hub
 from stable_baselines3 import PPO
 
-# Define a PPO model with MLP policy network
-model = PPO("MlpPolicy", "CartPole-v1", verbose=1)
+env = make_atari_env('SpaceInvadersNoFrameskip-v4', n_envs=1, seed=0)
```
Comment (osanseviero) [inline README.md:96]: `package_to_hub` example above uses different code, which might make the users focus on that instead of the pushing part. I would use same code.

File: README.md [hunk 9]
```diff
@@ -71,3 +115,19 @@ push_to_hub(
...
 )
 ```
+### Case 3: I use Google Colab
+- You can use xvbf (virtual screen)
+```
+!apt-get install -y xvfb python-opengl > /dev/null 2>&1
+```
+- Just put your code inside a python file and run
+```
+!xvfb-run -s "-screen 0 1400x900x24" <your_python_file>
```
Comment (osanseviero) [inline README.md:125]: Do you always need to do that if you want the package thing?

File: huggingface_sb3/load_from_hub.py [hunk 10]
```diff
@@ -17,19 +17,19 @@ def load_from_hub(repo_id: str, filename: str) -> str:
...
-                                             library_name="huggingface-sb3",
-                                             library_version="1.0.7")
-    
+        # Get the model from the Hugging Face Hub
+        model_download_url = hf_hub_url(repo_id=repo_id,
+                                        filename=filename)
+        # Downloading and caching the model on your local disk
+        downloaded_model_file = cached_download(model_download_url,
+                                                library_name="huggingface-sb3",
+                                                library_version="1.0.7")
```
Comment (osanseviero) [inline huggingface_sb3/load_from_hub.py:26]: There is a new method called `hf_hub_download` which does `hf_hub_url` +`cached_download`, so you can simplify all above to

```
downloaded_file = hf_hub_download(
  repo_id=repo_id,
  filename=filename,
  library_name=...
  l
)
```
https://github.com/huggingface/huggingface_hub/blob/3e72bc876d72855445a3c204f483eb761eb91da5/src/huggingface_hub/file_download.py

File: huggingface_sb3/load_from_hub.py [hunk 11]
```diff
@@ -17,19 +17,19 @@ def load_from_hub(repo_id: str, filename: str) -> str:
...
+
     # TODO: I need to find a way to override the HTTPError with mine
     except HTTPError:
-      raise HTTPError(
-          f"This {filename} does not exist. Please verify \n"
-          f"1. That your repository exists https://huggingface.co/{repo_id} \n"
-          f"2. That the filename you want to retrieve is the correct one and is a .zip file"
-      )
+        raise HTTPError(
```
Comment (osanseviero) [inline huggingface_sb3/load_from_hub.py:30]: Is this really needed? What error would `hf_hub_download` give the user if it points to a wrong repo or a wrong file?

File: huggingface_sb3/push_to_hub.py [hunk 12]
```diff
@@ -7,30 +7,309 @@
...
+import gym
+import zipfile
+
+from subprocess import call
+
 README_TEMPLATE = """---
 tags:
 - deep-reinforcement-learning
 - reinforcement-learning
-- stable-baselines3
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:14]: Why was this removed?

File: huggingface_sb3/push_to_hub.py [hunk 13]
```diff
@@ -7,30 +7,309 @@
...
+
+    # Step 1: Unzip the model
+    with zipfile.ZipFile(Path(repo_local_path) / model, 'r') as zip_ref:
+        zip_ref.extractall(Path(repo_local_path) / unzipped_model_folder)
+
+    # Step 2: Get data (JSON containing infos) and read it
+    with open(Path.joinpath(repo_local_path, unzipped_model_folder, 'data')) as json_file:
+        data = json.load(json_file)
+        # Add system_info elements to our JSON
+        data["system_info"] = stable_baselines3.get_system_info(print_info=False)[0]
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:52]: Nice!

File: huggingface_sb3/push_to_hub.py [hunk 14]
```diff
@@ -7,30 +7,309 @@
...
+
+    # Step 3: Write our config.json file
+    with open(Path(repo_local_path) / 'config.json', 'w') as outfile:
+        json.dump(data, outfile)
+
+
+def _evaluate_agent(model, eval_env, n_eval_episodes, is_deterministic, repo_local_path):
+    """
+    Evaluate the agent using SB3 evaluate_policy method and create a results.json
+    :param model: name of the model zip file
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:62]: is this the zip file or the model object?

File: huggingface_sb3/push_to_hub.py [hunk 15]
```diff
@@ -7,30 +7,309 @@
...
+    :param eval_env: environment used to evaluate the agent
+    :param n_eval_episodes: number of evaluation episodes (by default: 10)
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:64]: is it really 10 by default? Is that from `evaluate_policy`?

File: huggingface_sb3/push_to_hub.py [hunk 16]
```diff
@@ -7,30 +7,309 @@
...
+    :param is_deterministic: use deterministic or stochastic actions
+    :param repo_local_path: path of the local repository
+    """
+    # Step 1: Evaluate the agent
+    mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes, is_deterministic)
+
+    # Step 2: Create json evaluation
+    evaluate_data = {
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:72]: Very nice too! :rocket:

File: huggingface_sb3/push_to_hub.py [hunk 17]
```diff
@@ -7,30 +7,309 @@
...
+        name_prefix="",
+    )
+
+    obs = env.reset()
+    env.reset()
+    try:
+        for _ in range(video_length + 1):
+            action, _ = model.predict(obs, deterministic=is_deterministic)
+            obs, _, _, _ = env.step(action)
+            # env.render()
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:121]: nit: remove

File: huggingface_sb3/push_to_hub.py [hunk 18]
```diff
@@ -7,30 +7,309 @@
...
+        # Convert the video with x264 codec
+        inp = "./test.mp4"
+        out = "replay.mp4"
+        os.system(f"ffmpeg -y -i {inp} -vcodec h264 {out}".format(inp, out))
+        """
+        call([ffmpeg,
+              '-i', "./replay.mp4",
+              '-vcodec', h264,
+              replay.mp4])
+        """
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:138]: nit: remove

File: huggingface_sb3/push_to_hub.py [hunk 19]
```diff
@@ -7,30 +7,309 @@
...
+        # Move the video
+        shutil.move(os.path.join("./", "replay.mp4"), os.path.join(repo_local_path, "replay.mp4"))
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:140]: Would it be possible to work with the video directly in repo_local_path from the start? The disadvantage of how it's done right now is that if the script breaks, then users end up with an extraneous video in their working directory.

File: huggingface_sb3/push_to_hub.py [hunk 20]
```diff
@@ -7,30 +7,309 @@
...
+    except KeyboardInterrupt:
+        pass
+    except:
+        # Add a message for video
+        print("We are unable to generate a replay of your agent")
+        print("Please send a message to thomas.simonini@huggingface.co")
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:146]: I would instead suggest to open an issue :smile:

Comment (osanseviero) [reply huggingface_sb3/push_to_hub.py:146]: I would also suggest to use `logging` instead of `print` here

File: huggingface_sb3/push_to_hub.py [hunk 21]
```diff
@@ -7,30 +7,309 @@
...
+    """
+
+    model_card = f"""
+---
+      tags:
+      - {env_id}
+      - deep-reinforcement-learning
+      - reinforcement-learning
+      - stable-baselines3
+---
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:162]: Why not use `README_TEMPLATE` instead for this?

File: huggingface_sb3/push_to_hub.py [hunk 22]
```diff
@@ -7,30 +7,309 @@
...
+    :param env_id: name of the environment
+    """
+    entry_point = gym.envs.registry.env_specs[env_id].entry_point
+    return "AtariEnv" in str(entry_point)
+
+
+def _generate_replay(model, eval_env, video_length, is_deterministic, repo_local_path):
+    """
+    Generate a replay video of the agent
+    :param model: name of the model
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:100]: Should this be called `model_name` for consistency?

File: huggingface_sb3/push_to_hub.py [hunk 23]
```diff
@@ -7,30 +7,309 @@
...
+        shutil.move(os.path.join("./", "replay.mp4"), os.path.join(repo_local_path, "replay.mp4"))
+    except KeyboardInterrupt:
+        pass
+    except:
+        # Add a message for video
+        print("We are unable to generate a replay of your agent")
+        print("Please send a message to thomas.simonini@huggingface.co")
+
+
+def select_tags(env_id):
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:149]: I think a better function name would be something like `generate_metadata`

File: huggingface_sb3/push_to_hub.py [hunk 24]
```diff
@@ -7,30 +7,309 @@
...
+  ## Usage (with Stable-baselines3)
+  TODO: Add your code
+  """
+
+    return model_card
+
+
+def _create_model_card(repo_dir: Path, generated_model_card):
+    """Creates a model card for the repository.
+    TODO: Add metrics to model-index
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:199]: If possible I would already do this in this PR. You can use https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/repocard.py `metadata_save` and `metadata_eval_result` that should make your life easier than handling all the stuff yourself.

File: huggingface_sb3/push_to_hub.py [hunk 25]
```diff
@@ -7,30 +7,309 @@
...
+      - It generates the model card
+      - It generates a replay video of the agent
+      - It pushes everything to the hub
+      This is a work in progress function, if it does not work, use push_to_hub method
+      :param model: trained model
+      :param model_name: name of the model zip file
+      :param model_architecture: name of the architecture of your model (DQN, PPO, A2C, SAC...)
+      :param env_id: name of the environment
+      :param eval_env: environment used to evaluate the agent
+      :param repo_id: repo_id: id of the model repository from the Hugging Face Hub
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:241]:
```suggestion
      :param repo_id: id of the model repository from the Hugging Face Hub
```

File: huggingface_sb3/push_to_hub.py [hunk 26]
```diff
@@ -7,30 +7,309 @@
...
+      :param commit_message: commit message
+      :param is_deterministic: use deterministic or stochastic actions (by default: True)
+      :param n_eval_episodes: number of evaluation episodes (by default: 10)
+      :param use_auth_token
+      :param local_repo_path: local repository path
+      :param video_length:
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:247]:
```suggestion
```

File: huggingface_sb3/push_to_hub.py [hunk 27]
```diff
@@ -7,30 +7,309 @@
...
+      :param video_length: length of the video (in timesteps)
+      """
+    print(
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:250]: Let's use logging instead of printing across the repo

File: huggingface_sb3/push_to_hub.py [hunk 28]
```diff
@@ -7,30 +7,309 @@
...
+        "This function will save, evaluate, generate a video of your agent, create a model card and push everything to the hub. It might take up to 1min. \n This is a work in progress: If you encounter a bug, please send a message to thomas.simonini@huggingface.co and use push_to_hub instead.")
+    huggingface_token = HfFolder.get_token()
+
+    temp = repo_id.split('/')
+    organization = temp[0]
+    repo_name = temp[1]
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:256]:
```suggestion
    organization, repo_name = repo_id.split('/')
```

File: huggingface_sb3/push_to_hub.py [hunk 29]
```diff
@@ -7,30 +7,309 @@
...
+
+    # Step 1: Save the model
+    saved_model = model.save(Path(repo_local_path) / model_name)
+
+    # We create two versions of the environment one for video generation and one for evaluation
+    replay_env = eval_env
+
+    # Wrap the eval_env around a Monitor
+    # eval_env = Monitor(eval_env)
+    # replay_env = Monitor(replay_env)
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:285]: Should this be removed?

File: huggingface_sb3/push_to_hub.py [hunk 30]
```diff
@@ -7,30 +7,309 @@
...
+
+    # Deterministic by default (except for Atari)
+    is_deterministic = not is_atari(env_id)
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:288]: I think this will always override the `is_deterministic` param

File: huggingface_sb3/push_to_hub.py [hunk 31]
```diff
@@ -7,30 +7,309 @@
...
+
+    model_card += """
+  ## Usage (with Stable-baselines3)
+  TODO: Add your code
+  """
+
+    return model_card
+
+
+def _create_model_card(repo_dir: Path, generated_model_card):
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:197]: Maybe this should be called `_save_model_card` instead of `create_model_card` so it's not confusing with `generate_model_card`

File: huggingface_sb3/push_to_hub.py [hunk 32]
```diff
@@ -7,30 +7,309 @@
...
+      :param is_deterministic: use deterministic or stochastic actions (by default: True)
+      :param n_eval_episodes: number of evaluation episodes (by default: 10)
+      :param use_auth_token
+      :param local_repo_path: local repository path
+      :param video_length:
+      :param video_length: length of the video (in timesteps)
+      """
+    print(
+        "This function will save, evaluate, generate a video of your agent, create a model card and push everything to the hub. It might take up to 1min. \n This is a work in progress: If you encounter a bug, please send a message to thomas.simonini@huggingface.co and use push_to_hub instead.")
+    huggingface_token = HfFolder.get_token()
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:252]: I don't think you need to get the token. All the functions you use get the token under the hood. Passing the token is useful if you plan to make unit tests

**Review by osanseviero — COMMENTED** [review]
This looks amazing!! :rocket: I left mostly nits and minor suggestions that should improve overall codebase, but this is a great start :rocket:

File: README.md (hunk 2 above)
Comment (simoninithomas) [reply README.md:63]: Thanks!

File: .idea/huggingface_sb3.iml (hunk 1 above)
Comment (simoninithomas) [reply .idea/huggingface_sb3.iml:1]: Done :heavy_check_mark:

File: README.md (hunk 7 above)
Comment (simoninithomas) [reply README.md:90]: :heavy_check_mark:

File: README.md (hunk 8 above)
Comment (simoninithomas) [reply README.md]: :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 12 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:14]: My bad :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 15 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:64]: No is from package_to_hub line 218 n_eval_episodes=10

File: huggingface_sb3/push_to_hub.py (hunk 21 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:157]: Because `{env_id}` is not in `README_TEMPLATE`

File: huggingface_sb3/push_to_hub.py (hunk 25 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:236]: :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 26 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:242]: :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 27 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:245]: :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 20 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:141]: :heavy_check_mark:

File: huggingface_sb3/load_from_hub.py (hunk 10 above)
Comment (simoninithomas) [reply huggingface_sb3/load_from_hub.py]: ✔️

File: huggingface_sb3/push_to_hub.py (hunk 31 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py]: ✔️

File: huggingface_sb3/push_to_hub.py (hunk 32 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:244]: I was checking on other implementations but are you sure we can remove it?

Comment (osanseviero) [reply huggingface_sb3/push_to_hub.py:244]: `create_repo` accepts `token=None`, which is totally fine since when that happens it will just retrieve the token saved in the system with `HfFolder.get_token()`, which you're trying to do here.

I think what you likely want is something such as https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/fastai_utils.py#L352

So instead of `use_auth_token`, you would have an optional `token` param and pass it to `create_repo`. This is useful for colabs in which users are passing it, or for tests.

File: huggingface_sb3/load_from_hub.py (hunk 11 above)
Comment (simoninithomas) [reply huggingface_sb3/load_from_hub.py:28]: For now this part does nothing because I didn't overwrite the http error message of hf_hub_download.

File: huggingface_sb3/push_to_hub.py (hunk 32 above)
Comment (simoninithomas) [reply huggingface_sb3/push_to_hub.py:244]: :heavy_check_mark:

File: huggingface_sb3/push_to_hub.py (hunk 15 above)
Comment (osanseviero) [reply huggingface_sb3/push_to_hub.py:68]: I would remove the "(by default: 10)" from here, since the default is in the function that calls this function, not here.

File: huggingface_sb3/push_to_hub.py [hunk 33]
```diff
@@ -16,21 +24,330 @@
...
+    :std_reward: standard deviation of the mean reward of the agent
+    """
+    metadata = {}
+    metadata["library_name"] = "stable-baselines3"
+    metadata["tags"] = [
+        env_id,
+        "deep-reinforcement-learning",
+        "reinforcement-learning",
+        "stable-baselines3",
+    ]
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:175]: Super clean!

File: huggingface_sb3/push_to_hub.py [hunk 34]
```diff
@@ -41,56 +358,63 @@ def _copy_file(filepath: Path, dst_directory: Path):
...
 
-    temp = repo_id.split('/')
+    if token is None:
+        raise ValueError(
+            "You must login to the Hugging Face Hub. There are two options: "
+            "(1) Type `huggingface-cli login` in your terminal and enter your token. "
+            "(2) Enter your token in the `token` argument. "
+            "Your token is available in the Settings of your Hugging Face account. "
+        )
+
```
Comment (osanseviero) [inline huggingface_sb3/push_to_hub.py:386]: I think this can be removed. The downstream functions do a similar check to this.

**Review by osanseviero — APPROVED** [review]
Awesome!! Thanks for this, it's super cool! :rocket:

File: huggingface_sb3/push_to_hub.py [hunk 35]
```diff
@@ -1,11 +1,27 @@
-import logging
 import os
+import datetime
+
+import wasabi
```
[the remark this answers is not in the document]
Comment (osanseviero) [reply huggingface_sb3/push_to_hub.py]: Obsolete
